URL_CLINVAR_SINGLE = https://raw.githubusercontent.com/macarthur-lab/clinvar/master/output/b37/single/clinvar_allele_trait_pairs.single.b37.tsv.gz
URL_CLINVAR_MULTI = https://raw.githubusercontent.com/macarthur-lab/clinvar/master/output/b37/multi/clinvar_allele_trait_pairs.multi.b37.tsv.gz
URL_OMIM = ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/mim2gene_medgen
URL_HGNC = ftp://ftp.ebi.ac.uk/pub/databases/genenames/new/tsv/hgnc_complete_set.txt
URL_HPO = http://compbio.charite.de/jenkins/job/hpo.annotations.2018/lastSuccessfulBuild/artifact/misc_2018/phenotype.hpoa
URL_EXAC = ftp://ftp.broadinstitute.org/pub/ExAC_release/release0.3.1/ExAC.r0.3.1.sites.vep.vcf.gz
URL_GNOMAD_E = https://storage.googleapis.com/gnomad-public/release/2.0.2/vcf/exomes/gnomad.exomes.r2.0.2.sites.vcf.bgz
URL_GNOMAD_G1 = https://storage.googleapis.com/gnomad-public/release/2.0.2/vcf/genomes/gnomad.genomes.r2.0.2.sites.chr1.vcf.bgz
URL_GNOMAD_G2 = https://storage.googleapis.com/gnomad-public/release/2.0.2/vcf/genomes/gnomad.genomes.r2.0.2.sites.chr2.vcf.bgz
URL_GNOMAD_G3 = https://storage.googleapis.com/gnomad-public/release/2.0.2/vcf/genomes/gnomad.genomes.r2.0.2.sites.chr3.vcf.bgz
URL_GNOMAD_G4 = https://storage.googleapis.com/gnomad-public/release/2.0.2/vcf/genomes/gnomad.genomes.r2.0.2.sites.chr4.vcf.bgz
URL_GNOMAD_G5 = https://storage.googleapis.com/gnomad-public/release/2.0.2/vcf/genomes/gnomad.genomes.r2.0.2.sites.chr5.vcf.bgz
URL_GNOMAD_G6 = https://storage.googleapis.com/gnomad-public/release/2.0.2/vcf/genomes/gnomad.genomes.r2.0.2.sites.chr6.vcf.bgz
URL_GNOMAD_G7 = https://storage.googleapis.com/gnomad-public/release/2.0.2/vcf/genomes/gnomad.genomes.r2.0.2.sites.chr7.vcf.bgz
URL_GNOMAD_G8 = https://storage.googleapis.com/gnomad-public/release/2.0.2/vcf/genomes/gnomad.genomes.r2.0.2.sites.chr8.vcf.bgz
URL_GNOMAD_G9 = https://storage.googleapis.com/gnomad-public/release/2.0.2/vcf/genomes/gnomad.genomes.r2.0.2.sites.chr9.vcf.bgz
URL_GNOMAD_G10 = https://storage.googleapis.com/gnomad-public/release/2.0.2/vcf/genomes/gnomad.genomes.r2.0.2.sites.chr10.vcf.bgz
URL_GNOMAD_G11 = https://storage.googleapis.com/gnomad-public/release/2.0.2/vcf/genomes/gnomad.genomes.r2.0.2.sites.chr11.vcf.bgz
URL_GNOMAD_G12 = https://storage.googleapis.com/gnomad-public/release/2.0.2/vcf/genomes/gnomad.genomes.r2.0.2.sites.chr12.vcf.bgz
URL_GNOMAD_G13 = https://storage.googleapis.com/gnomad-public/release/2.0.2/vcf/genomes/gnomad.genomes.r2.0.2.sites.chr13.vcf.bgz
URL_GNOMAD_G14 = https://storage.googleapis.com/gnomad-public/release/2.0.2/vcf/genomes/gnomad.genomes.r2.0.2.sites.chr14.vcf.bgz
URL_GNOMAD_G15 = https://storage.googleapis.com/gnomad-public/release/2.0.2/vcf/genomes/gnomad.genomes.r2.0.2.sites.chr15.vcf.bgz
URL_GNOMAD_G16 = https://storage.googleapis.com/gnomad-public/release/2.0.2/vcf/genomes/gnomad.genomes.r2.0.2.sites.chr16.vcf.bgz
URL_GNOMAD_G17 = https://storage.googleapis.com/gnomad-public/release/2.0.2/vcf/genomes/gnomad.genomes.r2.0.2.sites.chr17.vcf.bgz
URL_GNOMAD_G18 = https://storage.googleapis.com/gnomad-public/release/2.0.2/vcf/genomes/gnomad.genomes.r2.0.2.sites.chr18.vcf.bgz
URL_GNOMAD_G19 = https://storage.googleapis.com/gnomad-public/release/2.0.2/vcf/genomes/gnomad.genomes.r2.0.2.sites.chr19.vcf.bgz
URL_GNOMAD_G20 = https://storage.googleapis.com/gnomad-public/release/2.0.2/vcf/genomes/gnomad.genomes.r2.0.2.sites.chr20.vcf.bgz
URL_GNOMAD_G21 = https://storage.googleapis.com/gnomad-public/release/2.0.2/vcf/genomes/gnomad.genomes.r2.0.2.sites.chr21.vcf.bgz
URL_GNOMAD_G22 = https://storage.googleapis.com/gnomad-public/release/2.0.2/vcf/genomes/gnomad.genomes.r2.0.2.sites.chr22.vcf.bgz
URL_GNOMAD_GX = https://storage.googleapis.com/gnomad-public/release/2.0.2/vcf/genomes/gnomad.genomes.r2.0.2.sites.chrX.vcf.bgz
URL_1000G1 = ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr1.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz
URL_1000G2 = ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr2.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz
URL_1000G3 = ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr3.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz
URL_1000G4 = ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr4.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz
URL_1000G5 = ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr5.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz
URL_1000G6 = ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr6.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz
URL_1000G7 = ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr7.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz
URL_1000G8 = ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr8.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz
URL_1000G9 = ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr9.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz
URL_1000G10 = ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr10.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz
URL_1000G11 = ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr11.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz
URL_1000G12 = ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr12.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz
URL_1000G13 = ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr13.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz
URL_1000G14 = ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr14.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz
URL_1000G15 = ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr15.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz
URL_1000G16 = ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr16.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz
URL_1000G17 = ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr17.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz
URL_1000G18 = ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr18.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz
URL_1000G19 = ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr19.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz
URL_1000G20 = ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr20.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz
URL_1000G21 = ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr21.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz
URL_1000G22 = ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr22.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz
URL_1000GX = ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chrX.phase3_shapeit2_mvncall_integrated_v1b.20130502.genotypes.vcf.gz
URL_DBSNP = ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh37p13/VCF/All_20180423.vcf.gz
URL_REF = ftp://ftp.ncbi.nlm.nih.gov/1000genomes/ftp/technical/reference/phase2_reference_assembly_sequence/hs37d5.fa.gz
URL_REFSEQ = ftp://ftp.ncbi.nlm.nih.gov/genomes/H_sapiens/ARCHIVE/ANNOTATION_RELEASE.105/GFF/ref_GRCh37.p13_top_level.gff3.gz
URL_ENSEMBL = ftp://ftp.ensembl.org/pub/release-75/gtf/homo_sapiens/Homo_sapiens.GRCh37.75.gtf.gz
URL_KNOWNGENEAA = http://hgdownload.cse.ucsc.edu/goldenpath/hg19/multiz100way/alignments/knownGene.exonAA.fa.gz

CLINVAR_SINGLE = ../databases/clinvar/download/$(notdir $(URL_CLINVAR_SINGLE))
CLINVAR_MULTI = ../databases/clinvar/download/$(notdir $(URL_CLINVAR_MULTI))
OMIM = ../databases/omim/download/$(notdir $(URL_OMIM))
HGNC = ../databases/hgnc/download/$(notdir $(URL_HGNC))
HPO = ../databases/hpo/download/$(notdir $(URL_HPO))
EXAC = ../databases/exac/download/$(notdir $(URL_EXAC))
GNOMAD_E = ../databases/gnomad_exomes/download/$(notdir $(URL_GNOMAD_E))
GNOMAD_G1 = ../databases/gnomad_genomes/download/$(notdir $(URL_GNOMAD_G1))
GNOMAD_G2 = ../databases/gnomad_genomes/download/$(notdir $(URL_GNOMAD_G2))
GNOMAD_G3 = ../databases/gnomad_genomes/download/$(notdir $(URL_GNOMAD_G3))
GNOMAD_G4 = ../databases/gnomad_genomes/download/$(notdir $(URL_GNOMAD_G4))
GNOMAD_G5 = ../databases/gnomad_genomes/download/$(notdir $(URL_GNOMAD_G5))
GNOMAD_G6 = ../databases/gnomad_genomes/download/$(notdir $(URL_GNOMAD_G6))
GNOMAD_G7 = ../databases/gnomad_genomes/download/$(notdir $(URL_GNOMAD_G7))
GNOMAD_G8 = ../databases/gnomad_genomes/download/$(notdir $(URL_GNOMAD_G8))
GNOMAD_G9 = ../databases/gnomad_genomes/download/$(notdir $(URL_GNOMAD_G9))
GNOMAD_G10 = ../databases/gnomad_genomes/download/$(notdir $(URL_GNOMAD_G10))
GNOMAD_G11 = ../databases/gnomad_genomes/download/$(notdir $(URL_GNOMAD_G11))
GNOMAD_G12 = ../databases/gnomad_genomes/download/$(notdir $(URL_GNOMAD_G12))
GNOMAD_G13 = ../databases/gnomad_genomes/download/$(notdir $(URL_GNOMAD_G13))
GNOMAD_G14 = ../databases/gnomad_genomes/download/$(notdir $(URL_GNOMAD_G14))
GNOMAD_G15 = ../databases/gnomad_genomes/download/$(notdir $(URL_GNOMAD_G15))
GNOMAD_G16 = ../databases/gnomad_genomes/download/$(notdir $(URL_GNOMAD_G16))
GNOMAD_G17 = ../databases/gnomad_genomes/download/$(notdir $(URL_GNOMAD_G17))
GNOMAD_G18 = ../databases/gnomad_genomes/download/$(notdir $(URL_GNOMAD_G18))
GNOMAD_G19 = ../databases/gnomad_genomes/download/$(notdir $(URL_GNOMAD_G19))
GNOMAD_G20 = ../databases/gnomad_genomes/download/$(notdir $(URL_GNOMAD_G20))
GNOMAD_G21 = ../databases/gnomad_genomes/download/$(notdir $(URL_GNOMAD_G21))
GNOMAD_G22 = ../databases/gnomad_genomes/download/$(notdir $(URL_GNOMAD_G22))
GNOMAD_GX = ../databases/gnomad_genomes/download/$(notdir $(URL_GNOMAD_GX))
1000G1 = ../databases/thousand_genomes/download/$(notdir $(URL_1000G1))
1000G2 = ../databases/thousand_genomes/download/$(notdir $(URL_1000G2))
1000G3 = ../databases/thousand_genomes/download/$(notdir $(URL_1000G3))
1000G4 = ../databases/thousand_genomes/download/$(notdir $(URL_1000G4))
1000G5 = ../databases/thousand_genomes/download/$(notdir $(URL_1000G5))
1000G6 = ../databases/thousand_genomes/download/$(notdir $(URL_1000G6))
1000G7 = ../databases/thousand_genomes/download/$(notdir $(URL_1000G7))
1000G8 = ../databases/thousand_genomes/download/$(notdir $(URL_1000G8))
1000G9 = ../databases/thousand_genomes/download/$(notdir $(URL_1000G9))
1000G10 = ../databases/thousand_genomes/download/$(notdir $(URL_1000G10))
1000G11 = ../databases/thousand_genomes/download/$(notdir $(URL_1000G11))
1000G12 = ../databases/thousand_genomes/download/$(notdir $(URL_1000G12))
1000G13 = ../databases/thousand_genomes/download/$(notdir $(URL_1000G13))
1000G14 = ../databases/thousand_genomes/download/$(notdir $(URL_1000G14))
1000G15 = ../databases/thousand_genomes/download/$(notdir $(URL_1000G15))
1000G16 = ../databases/thousand_genomes/download/$(notdir $(URL_1000G16))
1000G17 = ../databases/thousand_genomes/download/$(notdir $(URL_1000G17))
1000G18 = ../databases/thousand_genomes/download/$(notdir $(URL_1000G18))
1000G19 = ../databases/thousand_genomes/download/$(notdir $(URL_1000G19))
1000G20 = ../databases/thousand_genomes/download/$(notdir $(URL_1000G20))
1000G21 = ../databases/thousand_genomes/download/$(notdir $(URL_1000G21))
1000G22 = ../databases/thousand_genomes/download/$(notdir $(URL_1000G22))
1000GX = ../databases/thousand_genomes/download/$(notdir $(URL_1000GX))
DBSNP = ../databases/dbsnp/download/$(notdir $(URL_DBSNP))
REFSEQ = ../helpers/data/refseq/download/$(notdir $(URL_REFSEQ))
ENSEMBL = ../helpers/data/ensembl/download/$(notdir $(URL_ENSEMBL))
REF = ../helpers/data/reference/$(patsubst %.gz,%,$(notdir $(URL_REF)))
KNOWNGENEAA = ../databases/knowngeneaa/download/$(notdir $(URL_KNOWNGENEAA))


clinvar: $(CLINVAR_SINGLE) $(CLINVAR_MULTI)
$(CLINVAR_SINGLE):
	wget -c $(URL_CLINVAR_SINGLE) -O $@
$(CLINVAR_MULTI):
	wget -c $(URL_CLINVAR_MULTI) -O $@

omim: $(OMIM)
$(OMIM):
	wget -c $(URL_OMIM) -O $@

hgnc: $(HGNC)
$(HGNC):
	wget -c $(URL_HGNC) -O $@

hpo: $(HPO)
$(HPO):
	wget -c $(URL_HPO) -O $@

exac: $(EXAC)
$(EXAC):
	wget -c $(URL_EXAC) -O $@
	wget -c $(URL_EXAC).tbi -O $@.tbi

gnomad_exomes: $(GNOMAD_E)
$(GNOMAD_E):
	wget -c $(URL_GNOMAD_E) -O $@
	wget -c $(URL_GNOMAD_E).tbi -O $@.tbi

gnomad_genomes: $(GNOMAD_G1) $(GNOMAD_G2) $(GNOMAD_G3) $(GNOMAD_G4) $(GNOMAD_G5) $(GNOMAD_G6) $(GNOMAD_G7) $(GNOMAD_G8) $(GNOMAD_G9) $(GNOMAD_G10) $(GNOMAD_G11) $(GNOMAD_G12) $(GNOMAD_G13) $(GNOMAD_G14) $(GNOMAD_G15) $(GNOMAD_G16) $(GNOMAD_G17) $(GNOMAD_G18) $(GNOMAD_G19) $(GNOMAD_G20) $(GNOMAD_G21) $(GNOMAD_G22) $(GNOMAD_GX)
$(GNOMAD_G1):
	wget -c $(URL_GNOMAD_G1) -O $@
	wget -c $(URL_GNOMAD_G1).tbi -O $@.tbi
$(GNOMAD_G2):
	wget -c $(URL_GNOMAD_G2) -O $@
	wget -c $(URL_GNOMAD_G2).tbi -O $@.tbi
$(GNOMAD_G3):
	wget -c $(URL_GNOMAD_G3) -O $@
	wget -c $(URL_GNOMAD_G3).tbi -O $@.tbi
$(GNOMAD_G4):
	wget -c $(URL_GNOMAD_G4) -O $@
	wget -c $(URL_GNOMAD_G4).tbi -O $@.tbi
$(GNOMAD_G5):
	wget -c $(URL_GNOMAD_G5) -O $@
	wget -c $(URL_GNOMAD_G5).tbi -O $@.tbi
$(GNOMAD_G6):
	wget -c $(URL_GNOMAD_G6) -O $@
	wget -c $(URL_GNOMAD_G6).tbi -O $@.tbi
$(GNOMAD_G7):
	wget -c $(URL_GNOMAD_G7) -O $@
	wget -c $(URL_GNOMAD_G7).tbi -O $@.tbi
$(GNOMAD_G8):
	wget -c $(URL_GNOMAD_G8) -O $@
	wget -c $(URL_GNOMAD_G8).tbi -O $@.tbi
$(GNOMAD_G9):
	wget -c $(URL_GNOMAD_G9) -O $@
	wget -c $(URL_GNOMAD_G9).tbi -O $@.tbi
$(GNOMAD_G10):
	wget -c $(URL_GNOMAD_G10) -O $@
	wget -c $(URL_GNOMAD_G10).tbi -O $@.tbi
$(GNOMAD_G11):
	wget -c $(URL_GNOMAD_G11) -O $@
	wget -c $(URL_GNOMAD_G11).tbi -O $@.tbi
$(GNOMAD_G12):
	wget -c $(URL_GNOMAD_G12) -O $@
	wget -c $(URL_GNOMAD_G12).tbi -O $@.tbi
$(GNOMAD_G13):
	wget -c $(URL_GNOMAD_G13) -O $@
	wget -c $(URL_GNOMAD_G13).tbi -O $@.tbi
$(GNOMAD_G14):
	wget -c $(URL_GNOMAD_G14) -O $@
	wget -c $(URL_GNOMAD_G14).tbi -O $@.tbi
$(GNOMAD_G15):
	wget -c $(URL_GNOMAD_G15) -O $@
	wget -c $(URL_GNOMAD_G15).tbi -O $@.tbi
$(GNOMAD_G16):
	wget -c $(URL_GNOMAD_G16) -O $@
	wget -c $(URL_GNOMAD_G16).tbi -O $@.tbi
$(GNOMAD_G17):
	wget -c $(URL_GNOMAD_G17) -O $@
	wget -c $(URL_GNOMAD_G17).tbi -O $@.tbi
$(GNOMAD_G18):
	wget -c $(URL_GNOMAD_G18) -O $@
	wget -c $(URL_GNOMAD_G18).tbi -O $@.tbi
$(GNOMAD_G19):
	wget -c $(URL_GNOMAD_G19) -O $@
	wget -c $(URL_GNOMAD_G19).tbi -O $@.tbi
$(GNOMAD_G20):
	wget -c $(URL_GNOMAD_G20) -O $@
	wget -c $(URL_GNOMAD_G20).tbi -O $@.tbi
$(GNOMAD_G21):
	wget -c $(URL_GNOMAD_G21) -O $@
	wget -c $(URL_GNOMAD_G21).tbi -O $@.tbi
$(GNOMAD_G22):
	wget -c $(URL_GNOMAD_G22) -O $@
	wget -c $(URL_GNOMAD_G22).tbi -O $@.tbi
$(GNOMAD_GX):
	wget -c $(URL_GNOMAD_GX) -O $@
	wget -c $(URL_GNOMAD_GX).tbi -O $@.tbi

thousand_genomes: $(1000G1) $(1000G2) $(1000G3) $(1000G4) $(1000G5) $(1000G6) $(1000G7) $(1000G8) $(1000G9) $(1000G10) $(1000G11) $(1000G12) $(1000G13) $(1000G14) $(1000G15) $(1000G16) $(1000G17) $(1000G18) $(1000G19) $(1000G20) $(1000G21) $(1000G22) $(1000GX)
$(1000G1):
	wget -c $(URL_1000G1) -O $@
	wget -c $(URL_1000G1).tbi -O $@.tbi
$(1000G2):
	wget -c $(URL_1000G2) -O $@
	wget -c $(URL_1000G2).tbi -O $@.tbi
$(1000G3):
	wget -c $(URL_1000G3) -O $@
	wget -c $(URL_1000G3).tbi -O $@.tbi
$(1000G4):
	wget -c $(URL_1000G4) -O $@
	wget -c $(URL_1000G4).tbi -O $@.tbi
$(1000G5):
	wget -c $(URL_1000G5) -O $@
	wget -c $(URL_1000G5).tbi -O $@.tbi
$(1000G6):
	wget -c $(URL_1000G6) -O $@
	wget -c $(URL_1000G6).tbi -O $@.tbi
$(1000G7):
	wget -c $(URL_1000G7) -O $@
	wget -c $(URL_1000G7).tbi -O $@.tbi
$(1000G8):
	wget -c $(URL_1000G8) -O $@
	wget -c $(URL_1000G8).tbi -O $@.tbi
$(1000G9):
	wget -c $(URL_1000G9) -O $@
	wget -c $(URL_1000G9).tbi -O $@.tbi
$(1000G10):
	wget -c $(URL_1000G10) -O $@
	wget -c $(URL_1000G10).tbi -O $@.tbi
$(1000G11):
	wget -c $(URL_1000G11) -O $@
	wget -c $(URL_1000G11).tbi -O $@.tbi
$(1000G12):
	wget -c $(URL_1000G12) -O $@
	wget -c $(URL_1000G12).tbi -O $@.tbi
$(1000G13):
	wget -c $(URL_1000G13) -O $@
	wget -c $(URL_1000G13).tbi -O $@.tbi
$(1000G14):
	wget -c $(URL_1000G14) -O $@
	wget -c $(URL_1000G14).tbi -O $@.tbi
$(1000G15):
	wget -c $(URL_1000G15) -O $@
	wget -c $(URL_1000G15).tbi -O $@.tbi
$(1000G16):
	wget -c $(URL_1000G16) -O $@
	wget -c $(URL_1000G16).tbi -O $@.tbi
$(1000G17):
	wget -c $(URL_1000G17) -O $@
	wget -c $(URL_1000G17).tbi -O $@.tbi
$(1000G18):
	wget -c $(URL_1000G18) -O $@
	wget -c $(URL_1000G18).tbi -O $@.tbi
$(1000G19):
	wget -c $(URL_1000G19) -O $@
	wget -c $(URL_1000G19).tbi -O $@.tbi
$(1000G20):
	wget -c $(URL_1000G20) -O $@
	wget -c $(URL_1000G20).tbi -O $@.tbi
$(1000G21):
	wget -c $(URL_1000G21) -O $@
	wget -c $(URL_1000G21).tbi -O $@.tbi
$(1000G22):
	wget -c $(URL_1000G22) -O $@
	wget -c $(URL_1000G22).tbi -O $@.tbi
$(1000GX):
	wget -c $(URL_1000GX) -O $@
	wget -c $(URL_1000GX).tbi -O $@.tbi

dbsnp: $(DBSNP)
$(DBSNP):
	wget -c $(URL_DBSNP) -O $@
	wget -c $(URL_DBSNP).tbi -O $@.tbi

refseq: $(REFSEQ)
$(REFSEQ):
	wget -c $(URL_REFSEQ) -O $@

ensembl: $(ENSEMBL)
$(ENSEMBL):
	wget -c $(URL_ENSEMBL) -O $@

reference: $(REF)
$(REF):
	wget -c $(URL_REF) -O- \
	| gunzip -c > $@ || true

case:
	@echo -e
	@echo -e Place ped files \(yes, multiple are allowed\) in databases/case/download
	@echo -e

knowngeneaa: $(KNOWNGENEAA)
$(KNOWNGENEAA):
	wget -c $(URL_KNOWNGENEAA) -O $@

kegg:
	@echo -e
	@echo -e \#\# EnsemblToKegg
	@echo -e \* https://genome.ucsc.edu
	@echo -e \* Tools -\> Table Browser
	@echo -e \\t- group: All tables
	@echo -e \\t- assembly: GRCh37
	@echo -e \\t- table: keggPathway
	@echo -e \\t- output format: selected fields from primary and related tables
	@echo -e \\t- output file: genetokegg.tsv
	@echo -e \\t- get output
	@echo -e \\t\\t- Linked Tables -\> knownToEnsembl -\> allow selection from checked tables
	@echo -e \\t\\t- Linked Tables -\> ensGtp -\> allow selection from checked tables
	@echo -e \\t\\t- Select Fields from keggPathway -\> mapID
	@echo -e \\t\\t- ensGtp fields -\> gene
	@echo -e \\t\\t- get output
	@echo -e
	@echo -e \#\# RefseqToKegg
	@echo -e \* https://genome.ucsc.edu
	@echo -e \* Tools -\> Table Browser
	@echo -e \\t- group: All tables
	@echo -e \\t- assembly: GRCh37
	@echo -e \\t- table: keggPathway
	@echo -e \\t- output format: selected fields from primary and related tables
	@echo -e \\t- output file: genetokegg.tsv
	@echo -e \\t- get output
	@echo -e \\t\\t- Linked Tables -\> knownToKeggEntrez -\> allow selection from checked tables
	@echo -e \\t\\t- knownToKeggEntrez fields -\> keggEntrez
	@echo -e \\t\\t- get output
	@echo -e 
	@echo -e \#\# KeggInfo
	@echo -e \* https://genome.ucsc.edu
	@echo -e \* Tools -\> Table Browser
	@echo -e \\t- group: All tables
	@echo -e \\t- assembly: GRCh37
	@echo -e \\t- table: keggMapDesc
	@echo -e \\t- output format: selected fields from primary and related tables
	@echo -e \\t- output file: kegginfo.tsv
	@echo -e \\t- get output
	@echo -e \\t\\t- Select Fields from keggMapDesc -\> mapID
	@echo -e \\t\\t- Select Fields from keggMapDesc -\> description
	@echo -e \\t\\t- get output
	@echo -e
	@echo -e Save both files to databases/kegg/download
	@echo -e
